### ISSUE TITLE
STALE (Never Acknowledged) comes & ago every few seconds on all parameters on random portions of all pods

### ISSUE THREAD

**Issue by iftachsc** [istio/istio#13822]
After upgrading from a fully functional 1.0.6 installation, running istioctl proxy-status in endless loop will reveal more than half of the times many:  sidecars which are in 

```
CDS                            LDS                            EDS                                   RDS                            PILOT                            VERSION
STALE (Never Acknowledged)     STALE (Never Acknowledged)     STALE (Never Acknowledged) (20%)      SYNCED                         istio-pilot-5c457f6c79-nztp2     1.1.3 
```

here just pasting 1 our of like tens of them.
but then every 3-4 times i execute it  they will all be in synced.

we have more or less 400 injected pods. i've went through all incompetability changes from 1.0.6 in the docs. i even added a sidecar reoursce for each of the 3 injected namespaces.

issue still occurs.
pilots seems working significantly harder than in 1.0.6 - sometimes going to 3-5CPUs then dropping to nothing  (for 3-4 pilots)

![image](https://user-images.githubusercontent.com/28829551/57161730-f0bee180-6df4-11e9-8d86-6836e3454657.png)
and also:

![image](https://user-images.githubusercontent.com/28829551/57162154-113b6b80-6df6-11e9-84fc-b152ffac0417.png)

note: before i created sidecar resources (kind: sidecars) the same was just that the lower percentage numbers in the pic were mostly  100% and also some 50% - so i get since i did namespace segmentation it reduced the side car amount of configuration related it the pilots - correct me if i'm wrong here. i think i am wrng since some of the pods are on 100% in the same namespaces as other which have 20% and other numbers.

then if i keep running it, after few times i get synced state:

![image](https://user-images.githubusercontent.com/28829551/57161913-6aef6600-6df5-11e9-85fe-3ad15ba1250a.png)

but then the next time i run i see again the issue for several executions and then all again - synced state and so on. thus in most times the issue is seen. synced state is every now and then.
we also see many of the 
Omitting listener for management address .... due to collision with service listener  error
yes services' ports are the same as the deployments ports.. but we didnt see this in 1.0.6.

this results in netowrking issues we have between microservices.
rolling back to 1.0.6 using helm - made everything working again just as it was.

Importnat Note: running istio proxy-config endpoints shows that endpoints are healthy always. i.e. any proxy sees that others as healthy.

we also ran the following:
helm$ istioctl experimental verify-install --filename /tmp/installyay.yaml
Checked 0 crds
Checked 5 Istio Deployments
Istio is installed successfully

not sure it means anything. 

**Expected behavior**
to work great like we expected after reading 1.1.x improvements

**Steps to reproduce the bug**
not sure. we just upgrades according to the docs. i see many hit these issues separately. but not sure in this combination 

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
istioctl version:
client version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-20-gbc74657"}
galley version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}
ingressgateway version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-20-gbc74657"}
ingressgateway version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-20-gbc74657"}
ingressgateway version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-20-gbc74657"}
pilot version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}
pilot version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}
pilot version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}
sidecar-injector version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}
telemetry version: version.BuildInfo{Version:"1.1.4", GitRevision:"bc7465793cbff4c4189639b3f404e21c517cbdc6-dirty", User:"root", Host:"471e568b-66d7-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.3-20-gbc74657"}

kubectl version:
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.6-gke.10", GitCommit:"aaf0906400b5fc1d858ce0566a571e4f3ed06b9f", GitTreeState:"clean", BuildDate:"2019-03-30T19:30:48Z", GoVersion:"go1.10.8b4", Compiler:"gc", Platform:"linux/amd64"}

**How was Istio installed?**
upgraded using the 1.1.4 helm chart.
we've updated the values file according to the new installation options

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

additionally i ran the verify-install command:
*helm$ istioctl experimental verify-install --filename /tmp/installyay.yaml*
Checked 0 crds
Checked 5 Istio Deployments
Istio is installed successfully
**Affected product area (please put an X in all that apply)**

[x] Configuration Infrastructure
[ ] Docs
[x] Installation
[x] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by mandarjog** [comment]
Are you seeing any other problems like higher error rates associated with this observation?

**Comment by mandarjog** [comment]
@ostromart are we seeing this in 1.0.6 to 1.14 upgrades?

**Comment by iftachsc** [comment]
only what i wrote. we've seen here and there
Failed to create a new MCP sink stream: rpc error: code = Unavailable desc
but then we updated the installation without MCP. so it went away - the rest stayed.

in my lab i currently can restore the Never Acknowledged issue with our current gateway configuratin. we didnt change any istio yamls we had. we upgraded and thats it so it seem some gateway configuration that was ok for 1.0.6 isnt good for 1.1. but in my lab it is constant Never Acknowledged and in the real envirpnment it comes and goes.

**Comment by iftachsc** [comment]
i tried to do 1.0.6 -> 1.1.4 with bookinfo - allworks easily in the same way we upgrade the real environment. i am pretty sure this is related to some miss configuration but really there is no way to debug it or uinderstand. the upgrade notice has nothing but changing VirtualService/Gateways/DestRule etc changed needs to be done. there is only recommendation to move gateway resources to the same namespace as the gatwway workload it self. which we tried to do it (since curently ur gateway resources sites with the app namespace and its referrenced with a short name) and problem still persists..

**Comment by ostromart** [comment]
I've never seen this particular error during upgrade, but I haven't tried 1.1.4. I haven't looked at this issue fully but at first glance it seems like it's pilot related - let's chat offline @mandarjog .

**Comment by ostromart** [comment]
We suspect this might be a proxy-config tool reporting issue rather than a real underlying problem but let's look at both possibilities. Could you please run tools/dump_kubernetes.sh in your cluster?
This tool has some options. If istio (and nothing else) is installed to istio-system, running
  dump_kubernetes.sh -z -n istio-system
will give us the control plane logs.
We'd like to look at some proxy logs too. If you could label one or two pods and then run:
  dump_kubernetes.sh -z -n <pod namespace> -l <pod label>
That will produce proxy logs from the pods.
We normally just run the tool without parameters but your network is too large to capture all the logs without this filtering.

**Comment by iftachsc** [comment]
Ok. Since we rolled back the qa env I’ll do it ASAP.
Just to be clear: this is not the proxy just reporting an issue rather than a real issue.
As mentioned - in this time after the rollback there we networking issues. Proxies return 503 - also seen in kiali. Rolling istio + rolling the sidecars  back and everything went normal.

**Comment by ostromart** [comment]
I see. There's quite a lot of info in this issue, and it looks like at least a couple of likely unrelated problems:

1. After upgrade to 1.1.4, istioctl proxy-status reports sync problems
  Analysis: we suspect this may be an issue with the istioctl tool misreporting. If there's any other problem, it should show up in the control plane logs.
  Q: were any networking problems observed when cluster was in this state? i.e. before it was rolled back.

2. Networking problems during downgrade back to 1.0.6. Proxies report 503s.
  Need more info here - control plane and proxy logs will be a useful starting point.

**Comment by iftachsc** [comment]
answers by item
1. yes. after upgrading and before rollback there were networking issues. we atlest encoutnered in kiali one microsvc to microsvc path were we had 503 constantly.
and also pilots seem to work very hard. after rolling back to 1.0.6 it both were solved and system returned to normal.

2. i think you mis understood. there were no problem of networking in 1.0.6. only when we were with 1.1.4 proxies and 1.1.4 control plane.

anything else i can check?
ill provide the logs tomorrow. but i have 3 days windows to have the system in the bad state. i.e. if possible to get help during these 3 days. this thursday - sabbath. after that ill have to rollback again.

**Comment by ostromart** [comment]
Could someone from our networking team access your cluster to debug interactively? That would be the quickest way if you have no objection.

**Comment by iftachsc** [comment]
not directly i.e. cannot have credentials. possible through a session with me via hangouts or teamviewer or whatever. and yes id like that. it would be great.

**Comment by iftachsc** [comment]
I've upgraded the control plane - problem came back. i see it now even before rolling the sidecars. 
i am trying to run *dump_kubernetes.sh -z -n istio-system* and i get:

tools$ ./dump_kubernetes.sh -z -n istio-system
Fetching debug/configz from pilot
Error from server (NotFound): pods "istio-pilot-5b5c897786-58gdb istio-pilot-5b5c897786-6kthg istio-pilot-5b5c897786-mhthn istio-pilot-5b5c897786-tfpnr istio-pilot-5b5c897786-w5xw7 istio-pilot-5b5c897786-ztwgk" not found
Fetching debug/endpointz from pilot
Error from server (NotFound): pods "istio-pilot-5b5c897786-58gdb istio-pilot-5b5c897786-6kthg istio-pilot-5b5c897786-mhthn istio-pilot-5b5c897786-tfpnr istio-pilot-5b5c897786-w5xw7 istio-pilot-5b5c897786-ztwgk" not found
Fetching debug/adsz from pilot
Error from server (NotFound): pods "istio-pilot-5b5c897786-58gdb istio-pilot-5b5c897786-6kthg istio-pilot-5b5c897786-mhthn istio-pilot-5b5c897786-tfpnr istio-pilot-5b5c897786-w5xw7 istio-pilot-5b5c897786-ztwgk" not found
Fetching debug/authenticationz from pilot
Error from server (NotFound): pods "istio-pilot-5b5c897786-58gdb istio-pilot-5b5c897786-6kthg istio-pilot-5b5c897786-mhthn istio-pilot-5b5c897786-tfpnr istio-pilot-5b5c897786-w5xw7 istio-pilot-5b5c897786-ztwgk" not found
Fetching metrics from pilot
Error from server (NotFound): pods "istio-pilot-5b5c897786-58gdb istio-pilot-5b5c897786-6kthg istio-pilot-5b5c897786-mhthn istio-pilot-5b5c897786-tfpnr istio-pilot-5b5c897786-w5xw7 istio-pilot-5b5c897786-ztwgk" not found
Retrieving kubernetes resource configurations

---
just in case here is the pod list from this time
tools$ kubectl get pod -n istio-system

NAME                                      READY     STATUS    RESTARTS   AGE
istio-citadel-5c4f689c56-td4hj            1/1       Running   0          12m
istio-galley-79b86bbfcc-kqscw             1/1       Running   0          12m
istio-ingressgateway-7864ccc74f-lxrtw     1/1       Running   0          12m
istio-pilot-5b5c897786-58gdb              2/2       Running   0          11m
istio-pilot-5b5c897786-6kthg              2/2       Running   0          11m
istio-pilot-5b5c897786-mhthn              2/2       Running   0          12m
istio-pilot-5b5c897786-tfpnr              2/2       Running   0          11m
istio-pilot-5b5c897786-w5xw7              2/2       Running   0          11m
istio-pilot-5b5c897786-ztwgk              2/2       Running   0          10m
istio-policy-7f69f5c4d7-ppdnh             2/2       Running   0          12m
istio-sidecar-injector-7b58f755f7-b79k8   1/1       Running   0          12m
istio-telemetry-7c87b86488-6lhbt          2/2       Running   0          12m
prometheus-56c46f497d-ww7n4               1/1       Running   0          12m

**Comment by ostromart** [comment]
I've had some discussions with @mandarjog and someone from the networking team will work with you directly to debug the cluster (I work mainly on upgrade issues). btw. I did not get the logs but if you have them they may be useful for the folks doing the direct debugging.

**Comment by iftachsc** [comment]
Thanks.
I’ve sent you the logs via WeTransfer.
You should have them.
I know we have emery with mandar but I’m
Not sure about someone else’s from the net team.

Should I sent the logs to someone else?

**Comment by iftachsc** [comment]
update: we also see many of these:
	warn	ads	Failed to push, client busy sidecar~10.32.41.7~<podName>-665f9d6f58-

we have lots of them in istio pilot.

**Comment by mandarjog** [comment]
The logs show that one service in a bad state is responding with 404s to `/_/ping` health check requests.
Do you have any logging available at the application to see if this request is reaching it at all?

( service names and ip address changed )

```
[2019-05-10T14:12:49.411Z] "GET /_/ping HTTP/1.1" 503 UF "-" 0 91 0 - "-" "kube-probe/1.11+" "d4b07fb0-f338-4604-a78a-9ed529a82a8a" "10.10.9.7:30300" "127.0.0.1:30300" inbound|30300|http|sc-service.qa.svc.cluster.local - 10.10.9.7:30300 10.10.9.1:51018 -
[2019-05-10T14:07:10.586Z] "- - -" 0 - "-" 1474 546281 338718 - "-" "-" "-" "-" "10.10.39.16:3306" outbound|3306||mysql.qa.svc.cluster.local 10.10.9.7:44868 10.10.39.16:3306 10.10.9.7:44866 -
[2019-05-10T14:12:48.647Z] "GET /_/ping HTTP/1.1" 404 - "-" 0 48 2 2 "-" "kube-probe/1.11+" "d42892c3-2cb6-45af-8d07-d87af1a74047" "10.10.9.7:30300" "127.0.0.1:30300" inbound|30300|http|sc-service.qa.svc.cluster.local - 10.10.9.7:30300 10.10.9.1:51002 -
```

**Comment by mandarjog** [comment]
@costinm after an upgrade to 1.1.4 the customer is seeing pods not ready on two services specifically. The liveness check responds with a 404.

**Comment by howardjohn** [comment]
https://github.com/istio/istio/issues/14042 is at least partially responsible

**Comment by abhavan** [comment]
Has there been any further progress troubleshooting this issue? We’re seeing similar behavior after upgrading to 1.1.3 with just under 200 injected pods. Ref: https://discuss.istio.io/t/istioctl-proxy-status-keeps-running-stale-across-services-in-the-mesh-causing-sporadic-404s-for-requests-through-the-ingressgateway-s-blackhole-80-route/2377

**Comment by nbaechtold** [comment]
One other interesting observation that we have had is that HTTP routing config for the ingress pods seems to be flapping back and fourth between properly configured and configured only with a route for the blackhole config (that seems to be coming from here https://github.com/istio/istio/blob/release-1.1/pilot/pkg/networking/core/v1alpha3/gateway.go#L278).

```json
{
    "name": "http.80",
    "virtualHosts": [
        {
            "name": "blackhole:80",
            "domains": [
                "*"
            ],
            "routes": [
                {
                    "match": {
                        "prefix": "/"
                    },
                    "directResponse": {
                        "status": 404
                    },
                    "perFilterConfig": {
                        "mixer": {
                            "disable_check_calls": true
                        }
                    }
                }
            ]
        }
    ],
    "validateClusters": false
},
{
    "virtualHosts": [
        {
            "name": "backend",
            "domains": [
                "*"
            ],
            "routes": [
                {
                    "match": {
                        "prefix": "/stats/prometheus"
                    },
                    "route": {
                        "cluster": "prometheus_stats"
                    }
                }
            ]
        }
    ]
}
```

**Comment by wansuiye** [comment]
i met a similar problem, when the ingressgateway start(restart), the ingressgateway report logs at a random period of time(serveral seconds ~ minutes):
```
Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 0 rejected
```
meanwhile the pilot report:
```
2019-05-23T06:50:46.271046Z	info	ads	Pushing router~10.201.19.234~istio-ingressgateway-84c47b4849-vqx6g.istio-system~istio-system.svc.cluster.local-215
2019-05-23T06:50:46.271098Z	info	ads	Suppress push for router~10.201.19.234~istio-ingressgateway-84c47b4849-vqx6g.istio-system~istio-system.svc.cluster.local-215 at 2019-05-23T06:50:33Z/768, push with newer version 2019-05-23T06:50:46Z/770 in progress
2019-05-23T06:50:52.865950Z	warn	ads	Failed to push, client busy router~10.201.19.234~istio-ingressgateway-84c47b4849-vqx6g.istio-system~istio-system.svc.cluster.local-215
```
istio version:1.1.3
number of pod: 300

**Comment by vadimi** [comment]
We are having exactly the same issue in our cluster:

`Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 0 rejected`

and

`019-05-23T13:48:33.821006Z	warn	ads	Repeated failure to push sidecar~10.172.79.226~...`

istio `1.1.4`
number of pods around 350

**Comment by howardjohn** [comment]
Are all of your pilots overloaded (high CPU/memory)? The root cause of the original issue we determined was from pilot load due to https://github.com/istio/istio/issues/14042

**Comment by vadimi** [comment]
@howardjohn yes, pilot CPU is overloaded significantly.

**Comment by howardjohn** [comment]
> @howardjohn yes, pilot CPU is overloaded significantly.

Ok thanks, I think that is the cause. @mandarjog and I did some testing a while ago of Pilot under CPU throttling and found some issues but I don't think we ever found a root cause.

It would be useful to see why Pilot is loaded. When you say 99% CPU, what is your CPU limit on pilot?

How many pushes is it doing? Easiest way would be to query `sum(rate(pilot_xds_pushes{type!~".*_senderr"}[1m])) by (type)` on prometheus and post the graph

**Comment by vadimi** [comment]
I think I figured out why pilot is so loaded:

- we had some crashlooping pods in the cluster which was causing a lot of pilot CPU usage, I think #14042 is applicable here.

- what was causing 503 - is isio-proxy non graceful restarts in the application pod. Right before that it was reporting the following error: `PANIC: unprotected error in call to Lua API (27)` (issue #13722). We use envoy filters with lua scripts and as soon as I removed them 503 errors disappeared.

Basically in my case 503 were caused by crashing istio-proxy envoy instance. Even though 503 errors are fixed it still takes a lot if time for istio-proxy to get initial configuration from pilot, but there are other issues already opened for that.
